### PR TITLE
Prepare ark-snark for the new updates in ark-algebra

### PR DESCRIPTION
**Pull request by weikengchen**
## Description

This draft PR has several changes: (1) add patching, (2) change ark-test-curves feature to the curve one temporarily before a corresponding PR in ark-algebra is merged, and (3) fix a confusing "type annotations needed".

Since (2) is in a pending PR, it would be best not to merge this PR before that one is done, specifically giving that we are working on the `snark` repo. 

https://github.com/arkworks-rs/algebra/pull/461

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Re-reviewed `Files changed` in the Github PR explorer

N/A:
- [ ] Wrote unit tests
- [ ] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`

### CODE REVIEW

File: relations/src/r1cs/constraint_system.rs [hunk 1]
```diff
@@ -340,7 +341,11 @@ impl<F: Field> ConstraintSystem<F> {
                     let lc = transformed_lc_map
                         .get(&lc_index)
                         .expect("should be inlined");
-                    transformed_lc.extend((lc * coeff).0.into_iter());
```
Comment (Pratyush) [inline relations/src/r1cs/constraint_system.rs]: Why did this change?

**Comment by Pratyush** [comment]
This LGTM modulo the comment

File: relations/src/r1cs/constraint_system.rs (hunk 1 above)
Comment (findora-crypto) [reply relations/src/r1cs/constraint_system.rs]: In my local attempt it appears that "<&LinearCombination<F> as Mul<F>>::mul" is needed, otherwise the compiler is complaining. Actually, let me try again to see if this is still the case.

Comment (weikengchen) [reply relations/src/r1cs/constraint_system.rs]: A smaller change is made, so we continue to use `transformed_lc.extend((lc * coeff).0.into_iter());`

with the only change in signaling to the compiler the following:
```diff
- let mut transformed_lc_map = BTreeMap::new();
+ let mut transformed_lc_map = BTreeMap::<_, LinearCombination<F>>::new();
```

Comment (weikengchen) [reply relations/src/r1cs/constraint_system.rs]: Just found out "```diff"

Comment (weikengchen) [reply relations/src/r1cs/constraint_system.rs]: An approval will be needed.

File: snark/src/lib.rs [hunk 2]
```diff
@@ -23,7 +23,7 @@ pub trait SNARK<F: PrimeField> {
 
     /// The information required by the verifier to check a proof for a specific
     /// circuit *C*.
-    type VerifyingKey: Clone + ToBytes;
+    type VerifyingKey: Clone;
```
Comment (Pratyush) [inline snark/src/lib.rs]: Should we replace this with a `CanonicalSerialize + CanonicalDeserialize` bound?

Comment (Pratyush) [reply snark/src/lib.rs]: (Same for `ProvingKey`, `ProcessedVerifyingKey`, and `Proof`)

**Comment by weikengchen** [comment]
^ The change above (using CanonicalSerialize + CanonicalDeserialize) is done. But, it depends on https://github.com/arkworks-rs/algebra/pull/476

**Comment by weikengchen** [comment]
Closed --- needs to rebuild the PR.
